### PR TITLE
unixfs: add a directory interface

### LINKED ISSUE #5084: MFS Directory vs UnixFS/IO Directory
The `mfs` and the `unixfs.io` packages have both a `Directory` structure (the first containing the second) and both implementing an `AddChild` function (used in combination to add files/dirs to the MFS filesystem) which makes the code hard to understand when reading the MFS work flow (see https://github.com/ipfs/go-ipfs/issues/5081). It would seem that the rationale behind this is to have MFS representing an abstract directory and UnixFS providing the actual node to support that (but that is just my speculation).

The MFS `Directory` names its inner UnixFS/IO `Directory` as `dirbuilder` (without any comment) which makes it hard to understand their relationship besides what was previously mentioned,

https://github.com/ipfs/go-ipfs/blob/7853e53860805e08a212d78c4baa5d59bff99ba8/mfs/dir.go#L26-L41

https://github.com/ipfs/go-ipfs/blob/7853e53860805e08a212d78c4baa5d59bff99ba8/unixfs/io/dirbuilder.go#L28-L36

My main question right now is: what is the purpose of `unixfs.io.Directory`? If I understand correctly `unixfs.FSNode` is the UnixFS representaion for the directory, `mfs.Directory` is the same at the MFS layer, why is part of the UnixFS functionality regarding directories in a different structure (in `unixfs.io.Directory`)?

### PR DESCRIPTION
**Pull request by schomatis**
Addresses https://github.com/ipfs/go-ipfs/issues/5157.
Closes #5084.

----

```
unixfs: add a directory interface

Add a UnixFS `Directory` that hides implementation details and helps to
distinguish *what* is a UnixFS directory.

Replace the `unixfs.io.Directory` structure that contained the HAMT and basic
directory implementations (through inner pointers) with an interface containing
the same methods. Implement those methods in two clearly distinct structures for
each implementation (`BasicDirectory` and `HAMTDirectory`) avoiding pointer
logic and clearly differentiating which implementation does what.

The potential basic to HAMT transition was being hidden behind the `AddChild`
call at the UnixFS layer (changing one implementation pointer  with the other
one), it is now being explicitly done at the MFS layer.

Rename the `dirbuilder.go` file to `directory.go` and change the `Directory` MFS
attribute `dirbuilder` to `unixfsDir` to be consistent.
```

### CODE REVIEW

File: mfs/dir.go [hunk 1]
```diff
@@ -363,6 +363,18 @@ func (d *Directory) AddChild(name string, nd ipld.Node) error {
 		return err
 	}
 
+	//TODO: Resolve where to manage a possible transition to sharding.
```
Comment (schomatis) [inline mfs/dir.go]: This is ugly, and it's not equivalent to transitioning inside `unix.io.Directory.AddChild`.

File: mfs/dir.go [hunk 2]
```diff
@@ -372,6 +372,30 @@ func (d *Directory) AddChild(name string, nd ipld.Node) error {
 	return nil
 }
 
+// AddChildUnixFSDir adds a child to the inner UnixFS
+// directory and transitions to a HAMT sharding if needed.
+// TODO: Rename. Actually the real confusion is that
+// this structure has a `AddChild` function that calls
+// another `AddChild` function of its inner dir.
+func (d *Directory) AddChildUnixFSDir(name string, nd ipld.Node) error {
```
Comment (schomatis) [inline mfs/dir.go]: @whyrusleeping What do you think about moving the HAMT transition to the MFS layer?

Comment (whyrusleeping) [reply mfs/dir.go]: I guess the answer to that question depends on when we will need to do this, but I think its probably fine to have it in mfs

**Comment by schomatis** [comment]
[unixfs: add a directory interface](https://trello.com/c/pvXrREXR/5-unixfs-add-a-directory-interface)

**Comment by schomatis** [comment]
[unixfs: add a directory interface](https://trello.com/c/pvXrREXR/5-unixfs-add-a-directory-interface)

File: unixfs/io/directory.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	hamt "github.com/ipfs/go-ipfs/unixfs/hamt"
+
+	ipld "gx/ipfs/QmWi2BYBL5gJ3CiAiQchg6rn1A8iBsrWy51EYxvHVjFvLb/go-ipld-format"
+	cid "gx/ipfs/QmapdYm1b22Frv3k17fqrBYTFRxwiaVJkB299Mfn33edeB/go-cid"
+)
+
+// ShardSplitThreshold specifies how large of an unsharded directory
+// the Directory code will generate. Adding entries over this value will
+// result in the node being restructured into a sharded object.
+var ShardSplitThreshold = 1000
```
Comment (whyrusleeping) [inline unixfs/io/directory.go]: I think this is actually unused now. We should probably remove it

File: unixfs/io/directory.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+// Directory defines a UnixFS directory. It is used for creating, reading and
+// editing directories. It allows to work with different directory schemes,
+// like the basic or the HAMT implementation.
+//
+// It just allows to perform explicit edits on a single directory, working with
+// directory trees is out of its scope, they are managed by the MFS layer
+// (which is the main consumer of this interface).
+type Directory interface {
+
+	// SetPrefix sets the prefix of the root node.
```
Comment (whyrusleeping) [inline unixfs/io/directory.go]: i'd say "sets the cid prefix"

File: unixfs/io/directory.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+		return &HAMTDirectory{
+			dserv: dserv,
+			shard: shard,
+		}, nil
+	}
+
+	return nil, ErrNotADir
+}
+
+// SetPrefix implements the `Directory` interface.
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply unixfs/io/directory.go]: I think in go, for obvious interface implementation methods, they often just get left blank.  Saying 'implements X' is fine with me too

**Pull request merged**